### ISSUE TITLE
The training code for generation task seems exist bug.

### ISSUE THREAD

**Issue by codeview3d** [FoundationVision/UniTok#11]
Hi, thank you for your work. I carefully check the generation training code. I find that it can work for class_label to image generation. For the text-to-image generation, it cannot work. The reason as below:

for train_c2i.py
`x = x.to(device, non_blocking=True)

  y = y.to(device, non_blocking=True)

  z_indices = x.squeeze(dim=1)

  c_indices = y.reshape(-1)

  assert z_indices.shape[0] == c_indices.shape[0]

  with torch.cuda.amp.autocast(dtype=ptdtype):

         _, loss = model(cond_idx=c_indices, idx=z_indices[:, :, :-1], targets=z_indices)
`
we can see that the input idx is (B, K, L-1), K denotes the number of codebook. The c_indices will produce conditional embedding by LabelEmbedder to (B, 1, D). 
token_embeddings = torch.cat((cond_embeddings, token_embeddings), dim=1)
we can see that the token_embeddings can align with the target length (L). 

However, for the train_t2i.py:
`       _, loss = model(cond_idx=c_indices, idx=z_indices[:,:-1], targets=z_indices, mask=attn_mask[:, :, :-1,:-1], valid=valid)
`
the input index become the (B, L), without consider the K codebooks. Furthermore, the c_indices is (B, 120, D). 

token_embeddings = torch.cat((cond_embeddings, token_embeddings), dim=1) will produce (120+L-1) sequence length. Cannot align with target length L.

**Comment by machuofan** [comment]
Hi @codeview3d, thanks for pointing out this. We have only customized LlamaGen for class-conditional image generation with UniTok. To support text-to-image generation, you may need to modify corresponding files as we do in [extract_codes_c2i.py](https://github.com/FoundationVision/UniTok/blob/main/eval/generation/autoregressive/train/extract_codes_c2i.py),  [train_c2i.py](https://github.com/FoundationVision/UniTok/blob/main/eval/generation/autoregressive/train/train_c2i.py) and [sample_c2i_ddp.py](https://github.com/FoundationVision/UniTok/blob/main/eval/generation/autoregressive/sample/sample_c2i_ddp.py).

**Comment by Masaaki-75** [comment]
There seems still a potential bug in `train_c2i.py`. 

After the autoregressive head, the output `logits` should have shape `[B, 1+(K-1), 1+(Li-1), V/K]`, where `K` is the number of codebooks, `Li` is the number of image tokens, `V` is the overall vocabulary size.

The next line then reads (see here: https://github.com/FoundationVision/UniTok/blob/main/eval/llamagen/autoregressive/models/gpt.py#L464): 
`logits = logits[:, self.cls_token_num - 1:].contiguous()`

One thing confused me: why the indexing is taken on the dimension for `K`, instead of the dimension for `Li` (i.e., `logits[:, :, self.cls_token_num-1]`, note the dimension for indexing)?

That said, since the default value of `cls_token_num` is exactly 1, the current line has NO effect on the `logits`.

**Comment by machuofan** [comment]
> There seems still a potential bug in `train_c2i.py`.
> 
> After the autoregressive head, the output `logits` should have shape `[B, 1+(K-1), 1+(Li-1), V/K]`, where `K` is the number of codebooks, `Li` is the number of image tokens, `V` is the overall vocabulary size.
> 
> The next line then reads (see here: https://github.com/FoundationVision/UniTok/blob/main/eval/llamagen/autoregressive/models/gpt.py#L464): `logits = logits[:, self.cls_token_num - 1:].contiguous()`
> 
> One thing confused me: why the indexing is taken on the dimension for `K`, instead of the dimension for `Li` (i.e., `logits[:, :, self.cls_token_num-1]`, note the dimension for indexing)?
> 
> That said, since the default value of `cls_token_num` is exactly 1, the current line has NO effect on the `logits`.

Thanks for pointing out this potential bug. Yes, you are correct, it should be `logits[:, :, self.cls_token_num-1]`. Since cls_token_num is always 1 in class-conditional image generation, it does not make a difference here. But it may cause a bug in text-conditional image generation.
